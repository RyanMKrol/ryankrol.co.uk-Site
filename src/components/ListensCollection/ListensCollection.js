import React, { Component } from 'react';
import fetch from 'node-fetch';

import baseEndpoint from '../../utils/ProdUtils';
import styles from './ListensCollection.module.css';

import ListensItemFull from '../ListensItemFull';
import ListensItemMini from '../ListensItemMini';

const LISTENS_API_ENDPOINT = `${baseEndpoint()}/listens`;

/**
 * Method to get listens data
 *
 * @returns {Promise<JSON>} Listens data
 */
async function fetchData() {
  return fetch(LISTENS_API_ENDPOINT).then((response) => response.json());
}

/**
 * Internal class for rendering a header item
 */
class ListensCollection extends Component {
  /**
   * @class
   * @param {object} props Contains the desired output size of the result
   */
  constructor(props) {
    super();

    this.state = {
      data: null,
    };

    this.outputSize = props.outputSize;
  }

  /**
   * Fetches data after component creation
   */
  componentDidMount() {
    fetchData().then((data) => {
      this.setState({
        data,
      });
    });
  }

  /**
   * Renders the component
   *
   * @returns {React.Component} JSX for individual header link
   */
  render() {
    const collectionData = this.state.data
      ? this.state.data.map((item) => (this.outputSize === 'full' ? (
            <ListensItemFull
              albumLink={item.albumLink}
              albumName={item.albumName}
              thumbnail={item.thumbnail}
              playcount={item.playcount}
              artist={item.artist}
            />
      ) : (
            <ListensItemMini
              albumLink={item.albumLink}
              albumName={item.albumName}
              thumbnail={item.thumbnail}
            />
      )))
      : null;
    return (
      <div>
        <div className={styles.listensCollection}>{collectionData}</div>
      </div>
    );
  }
}

export default ListensCollection;
