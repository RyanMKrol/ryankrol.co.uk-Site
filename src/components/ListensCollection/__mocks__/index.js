import ListensCollection from '../ListensCollection';

const FULL_OUTPUT_MOCK = <ListensCollection outputSize="full" />;
const REDUCED_OUTPUT_MOCK = <ListensCollection outputSize="reduced" />;

const EXAMPLE_LISTENS_MOCK_ONE = {
  albumLink: 'test-albumLink-one',
  albumName: 'test-albumName-one',
  thumbnail: 'test-thumbnail-one',
  artist: 'test-artist-one',
  playcount: 1,
};

const EXAMPLE_LISTENS_MOCK_TWO = {
  albumLink: 'test-albumLink-two',
  albumName: 'test-albumName-two',
  thumbnail: 'test-thumbnail-two',
  artist: 'test-artist-two',
  playcount: 2,
};

export {
  FULL_OUTPUT_MOCK,
  REDUCED_OUTPUT_MOCK,
  EXAMPLE_LISTENS_MOCK_ONE,
  EXAMPLE_LISTENS_MOCK_TWO,
};
