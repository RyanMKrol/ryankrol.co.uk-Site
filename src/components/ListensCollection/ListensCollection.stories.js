import React from 'react';

import { FULL_OUTPUT_MOCK, REDUCED_OUTPUT_MOCK } from './__mocks__';

export default {
  title: 'Components/ListensCollection',
};

/**
 * story for the full output
 *
 * @returns {React.Component} JSX for the ListensCollection
 */
export const fullOutput = () => FULL_OUTPUT_MOCK;

/**
 * story for the reduced output
 *
 * @returns {React.Component} JSX for the ListensCollection
 */
export const reducedOutput = () => REDUCED_OUTPUT_MOCK;
