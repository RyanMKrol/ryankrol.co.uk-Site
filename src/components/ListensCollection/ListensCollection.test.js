import React from 'react';
import fetch from 'node-fetch';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import flushPromises from '../../utils/TestUtils';
import ListensCollection from './ListensCollection';
import {
  FULL_OUTPUT_MOCK,
  REDUCED_OUTPUT_MOCK,
  EXAMPLE_LISTENS_MOCK_ONE,
  EXAMPLE_LISTENS_MOCK_TWO,
} from './__mocks__';

jest.mock('node-fetch');

beforeEach(() => {
  // as this has been mocked once, i need to always be supplying a value, or the shallow fails
  fetch.mockResolvedValueOnce({
    json: () => [EXAMPLE_LISTENS_MOCK_ONE],
  });
});

it('renders ListensCollection with reduced output', async () => {
  const listensCollection = shallow(REDUCED_OUTPUT_MOCK);

  await flushPromises();

  expect(toJson(listensCollection)).toMatchSnapshot();
});

it('renders ListensCollection with full output', async () => {
  const listensCollection = shallow(FULL_OUTPUT_MOCK);

  await flushPromises();

  expect(toJson(listensCollection)).toMatchSnapshot();
});

it('renders ListensCollection with reduced output, with multiple values', async () => {
  fetch.mockResolvedValueOnce({
    json: () => [EXAMPLE_LISTENS_MOCK_ONE, EXAMPLE_LISTENS_MOCK_TWO],
  });
  const listensCollection = shallow(REDUCED_OUTPUT_MOCK);

  await flushPromises();

  expect(toJson(listensCollection)).toMatchSnapshot();
});

it('renders ListensCollection with full output, with multiple values', async () => {
  fetch.mockResolvedValueOnce({
    json: () => [EXAMPLE_LISTENS_MOCK_ONE, EXAMPLE_LISTENS_MOCK_TWO],
  });
  const listensCollection = shallow(FULL_OUTPUT_MOCK);

  await flushPromises();

  expect(toJson(listensCollection)).toMatchSnapshot();
});

it('ListensCollection does not throw with reduced output', async () => {
  expect(() => mount(REDUCED_OUTPUT_MOCK)).not.toThrow();
});

it('ListensCollection does not throw with full output', async () => {
  expect(() => mount(FULL_OUTPUT_MOCK)).not.toThrow();
});
