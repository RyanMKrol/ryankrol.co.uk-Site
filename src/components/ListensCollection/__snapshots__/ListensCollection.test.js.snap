// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ListensCollection with full output 1`] = `
<div>
  <div
    className="listensCollection"
  >
    <ListensItemFull
      albumLink="test-albumLink-one"
      albumName="test-albumName-one"
      artist="test-artist-one"
      playcount={1}
      thumbnail="test-thumbnail-one"
    />
  </div>
</div>
`;

exports[`renders ListensCollection with full output, with multiple values 1`] = `
<div>
  <div
    className="listensCollection"
  >
    <ListensItemFull
      albumLink="test-albumLink-one"
      albumName="test-albumName-one"
      artist="test-artist-one"
      playcount={1}
      thumbnail="test-thumbnail-one"
    />
  </div>
</div>
`;

exports[`renders ListensCollection with reduced output 1`] = `
<div>
  <div
    className="listensCollection"
  >
    <ListensItemMini
      albumLink="test-albumLink-one"
      albumName="test-albumName-one"
      thumbnail="test-thumbnail-one"
    />
  </div>
</div>
`;

exports[`renders ListensCollection with reduced output, with multiple values 1`] = `
<div>
  <div
    className="listensCollection"
  >
    <ListensItemMini
      albumLink="test-albumLink-one"
      albumName="test-albumName-one"
      thumbnail="test-thumbnail-one"
    />
  </div>
</div>
`;
